spark/v3.1/spark/src/main/java/com/netease/arctic/spark/SparkInternalRowCastWrapper.java

```diff
@@ -45,7 +46,96 @@ public SparkInternalRowCastWrapper(InternalRow row, StructType schema, ChangeAct
       this.changeAction = changeAction;
     }
     this.isDelete = isDelete;
-    this.isUpsert = true;
+    this.isMerge = false;
+  }
+
+  public SparkInternalRowCastWrapper(InternalRow row, ChangeAction changeAction) {
+    this.row = buildProjectingInternalRow((ProjectingInternalRow)row);
+    this.changeAction = changeAction;
+    this.schema = ((ProjectingInternalRow) row).schema();
+    this.isMerge = true;
+  }
+
+  private InternalRow buildProjectingInternalRow(ProjectingInternalRow projectingInternalRow) {
+    InternalRow row = projectingInternalRow.getRow();
+    StructType rowSchema = projectingInternalRow.schema();
+    List<DataType> dataTypeList = Arrays.stream(rowSchema.fields())
+        .map(StructField::dataType).collect(Collectors.toList());
+    List<Integer> colOrdinals = JavaConverters.seqAsJavaList(projectingInternalRow.colOrdinals())
+            .stream().map(o -> Integer.parseInt(o.toString())).collect(Collectors.toList());
+    List<Object> rows = new ArrayList<>();
+    for (int i = 0; i < dataTypeList.size(); i++) {
+      rows.add(row.get(colOrdinals.get(i), dataTypeList.get(i)));
+    }
+    return new GenericInternalRow(rows.toArray());
+  }
```

Comment:
Is this needed? need to check

Reply:
No, it is really unnecessary, I will remove